src/Processors/Transforms/DistinctTransform.cpp

```diff
@@ -48,6 +52,108 @@ void DistinctTransform::buildFilter(
     }
 }
 
+template <typename Method>
+void DistinctTransform::buildFilterWithMask(
+    Method & method,
+    const ColumnRawPtrs & columns,
+    IColumn::Filter & filter,
+    const size_t rows,
+    SetVariants & variants,
+    const IColumn::Filter & mask) const
+{
+    typename Method::State state(columns, key_sizes, nullptr);
+
+    for (size_t i = 0; i < rows; ++i)
+    {
+        if (!mask[i])
+        {
+            /// Already known duplicate row (by LC index), skip insertion
+            filter[i] = 0;
+            continue;
+        }
+
+        auto emplace_result = state.emplaceKey(method.data, i, variants.string_pool);
+        filter[i] = emplace_result.isInserted();
+    }
+}
+
+IColumn::Filter DistinctTransform::buildLowCardinalityMask(const ColumnLowCardinality & column, size_t num_rows)
+{
+    IColumn::Filter mask(num_rows, 0);
+
+    const auto & dictionary = column.getDictionary();
+    const auto dict_size = dictionary.size();
+
+    LCDictionaryKey dict_key;
+    dict_key.hash = dictionary.getHash();
```

Comment:
This seems aggressive. Do we need to hash the whole dictionary? Maybe it's shared and we can validate based on the pointer, which will be free